x/proof/keeper/session.go

```diff
@@ -133,6 +133,59 @@ func (k msgServer) validateClaimWindow(
 	return nil
 }
 
+// validateProofWindow returns an error if the given session is not eligible for proving.
+// It *assumes* that the msg's session header is a valid on-chain session with correct
+// height fields. First call #queryAndValidateSessionHeader to ensure any user-provided
+// session header is valid and correctly hydrated.
+func (k msgServer) validateProofWindow(
+	ctx context.Context,
+	msg *types.MsgSubmitProof,
+) error {
+	logger := k.Logger().With("method", "validateProofWindow")
+	sessionHeader := msg.GetSessionHeader()
+	sharedParams := k.sharedKeeper.GetParams(ctx)
+
+	sessionEndHeight := sessionHeader.GetSessionEndBlockHeight()
+
+	// Get the proof window open and close heights for the given session header.
+	proofWindowOpenHeight := shared.GetProofWindowOpenHeight(&sharedParams, sessionEndHeight)
+	proofWindowCloseHeight := shared.GetProofWindowCloseHeight(&sharedParams, sessionEndHeight)
+
+	// Get the current block height.
+	sdkCtx := cosmostypes.UnwrapSDKContext(ctx)
+	currentHeight := sdkCtx.BlockHeight()
+
+	// Ensure the current block height is AFTER the proof window open height.
+	if currentHeight < proofWindowOpenHeight {
+		return types.ErrProofProofOutsideOfWindow.Wrapf(
+			"current block height %d is less than session proof window open height %d",
```

Comment:
Optional (here and elsewhere). I found it easier to read while debugging

```suggestion
			"current block height (%d) is less than session proof window open height (%d)",
```